### PR TITLE
Tools API to send data message with custom parents

### LINKED ISSUE #2074: Tools API to post an entire message
## Description

We want to be able to control the selection of parents when constructing a message. This helps create more complicated scenarios.
Hence, message construction is delegated from the `MessageFactory` to the client.

- Create a `POST` endpoint `/tools/message` that accepts the message bytes.
- The handler parses the message and does some checks. e.g If the `nonce` is 0, perform PoW on the node and update the message.
- Then it passes the message to the normal gossip flow by triggering `MessageConstructedEvent`.
- The handler should wait for the message to be stored before returning success. Else, return an error.

### PR DESCRIPTION
**Pull request by acha-bill**
# Description of change

API to send a data message with custom parents.

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested
Added integration test

## Change checklist

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

Fixes #2074

### CODE REVIEW

File: plugins/webapi/tools/message/send_message.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	if len(request.Data) == 0 {
+		return c.JSON(http.StatusBadRequest, jsonmodels.DataResponse{Error: "no data provided"})
+	}
+
+	msg, _, err := tangle.MessageFromBytes(request.Data)
+	if err != nil {
+		return c.JSON(http.StatusBadRequest, jsonmodels.DataResponse{Error: fmt.Sprintf("error: %s", err.Error())})
+	}
+
+	if msg.Version() != tangle.MessageVersion {
```
Comment (daria305) [inline plugins/webapi/tools/message/send_message.go:41]: Why should user even care if he puts a correct version in the API request if the message is prepared by node anyway?

File: tools/integration-tests/tester/tests/diagnostics/diagnostics_test.go [hunk 2]
```diff
@@ -96,3 +102,42 @@ func TestDiagnosticApis(t *testing.T) {
...
+	api := node.GoShimmerAPI
+
+	sent := tests.SendDataMessages(t, n.Peers()[:1], 5)
+	parentMessageIDs := tangle.NewParentMessageIDs()
+
+	for k := range sent {
+		parentID, _ := tangle.NewMessageID(k)
+		parentMessageIDs = parentMessageIDs.Add(tangle.StrongParentType, parentID)
+	}
+	msg, err := tangle.NewMessage(
```
Comment (daria305) [inline tools/integration-tests/tester/tests/diagnostics/diagnostics_test.go:124]: Message is prepared anyway by the node we send the request to, it is a little misleading that we need to create this dummy message and put the bytes. What about passing only parents and payload to api.SendMessage() and this wrapping and changing to bytes will be done by client lib

File: plugins/webapi/tools/message/send_message.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	var parents tangle.MessageIDsSlice
+	msg.ForEachParent(func(parent tangle.Parent) {
+		parents = append(parents, parent.ID)
+	})
+	issuingTime := deps.Tangle.MessageFactory.GetIssuingTime(parents)
+	sequenceNumber, err := deps.Tangle.MessageFactory.NextSequenceNumber()
+	if err != nil {
+		return c.JSON(http.StatusBadRequest, jsonmodels.DataResponse{Error: err.Error()})
+	}
+	nonce, err := deps.Tangle.MessageFactory.DoPOW(references, issuingTime, issuerPublicKey, sequenceNumber, msg.Payload())
```
Comment (daria305) [inline plugins/webapi/tools/message/send_message.go:60]: Out of curiosity, have you considered the version with PoW done by the requester, not by the node? Probably this sequence number would be a problem

**Review by daria305 — CHANGES_REQUESTED** [review]
Looks good, I left some comments you might consider to add

File: plugins/webapi/tools/message/send_message.go (hunk 1 above)
Comment (acha-bill) [reply plugins/webapi/tools/message/send_message.go:41]: yeah. We shouldn't care about this.

File: plugins/webapi/tools/message/send_message.go (hunk 3 above)
Comment (acha-bill) [reply plugins/webapi/tools/message/send_message.go:60]: Not put any thought into this. :)

File: tools/integration-tests/tester/tests/diagnostics/diagnostics_test.go (hunk 2 above)
Comment (acha-bill) [reply tools/integration-tests/tester/tests/diagnostics/diagnostics_test.go:124]: Great idea actually. We should have the API only accept parents and payload. :)

File: plugins/webapi/tools/message/send_message.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	signature, err := deps.Tangle.MessageFactory.Sign(references, issuingTime, issuerPublicKey, sequenceNumber, msgPayload, nonce)
+	if err != nil {
+		return c.JSON(http.StatusBadRequest, jsonmodels.DataResponse{Error: err.Error()})
+	}
+	msg, err := tangle.NewMessage(references, issuingTime, issuerPublicKey, sequenceNumber, msgPayload, nonce, signature)
+	if err != nil {
+		return c.JSON(http.StatusBadRequest, jsonmodels.DataResponse{Error: err.Error()})
+	}
+
+	go deps.Tangle.MessageFactory.Events.MessageConstructed.Trigger(msg)
```
Comment (jkrvivian) [inline plugins/webapi/tools/message/send_message.go:75]: What about wrapping those in a function in the messagefactory?
Maybe name the function to `IssuePayloadWithParents`, then there's no need to export doPoW, sign and such.

**Review by daria305 — APPROVED** [review]

**Pull request merged**